security/webauthn/src/test/java/AbstractWebAuthnTest.java

```diff
@@ -0,0 +1,228 @@
+    private static final String LOGOUT_URL = "/q/webauthn/logout";
+
+    private static final String PUBLIC_API_URL = "/api/public";
+    private static final String PUBLIC_ME_API_URL = "/api/public/me";
+    private static final String USER_API_URL = "/api/users/me";
+    private static final String ADMIN_API_URL = "/api/admin";
+
+    private static final String USERNAME = "Roosvelt";
+
+    private static Filter cookieFilter;
+
+    enum User {
+        USER,
+        ADMIN;
+    }
+
+    @BeforeAll
+    public static void setup() {
+        cookieFilter = new CookieFilter();
+
+    }
+
+    @Test
+    @Order(1)
+    public void checkLogoutInitial() {
+        verifyLoggedOut(cookieFilter);
+    }
+
+    @Test
+    @Order(2)
+    public void checkAdminAPIWithoutUser() {
+        given()
+                .redirects().follow(false)
+                .get(ADMIN_API_URL)
+                .then()
+                .statusCode(302);
+    }
+
+    @Test
+    @Order(3)
+    public void checkMeAPIWithoutUser() {
+        given()
+                .redirects().follow(false)
+                .get(USER_API_URL)
+                .then()
+                .statusCode(302);
+    }
+
+    @Test
+    @Order(4)
+    public void checkPublicAPI() {
+        given()
+                .get(PUBLIC_API_URL)
+                .then()
+                .statusCode(200)
+                .body(Matchers.is("public"));
+    }
+
+    @Test
+    @Order(5)
+    public void testRegisterWebAuthnUser() {
+        MyWebAuthnHardware myWebAuthnHardware = new MyWebAuthnHardware();
+        String challenge = getChallenge(USERNAME, cookieFilter);
+        JsonObject registrationJson = myWebAuthnHardware.makeRegistrationJson(challenge);
+        invokeCallback(registrationJson, cookieFilter);
+        verifyLoggedIn(cookieFilter, USERNAME, User.USER);
+        invokeUserLogout();
+
+    }
+
+    @Test
+    @Order(6)
+    public void testRegisterSameUserName() {
+        MyWebAuthnHardware myWebAuthnHardware = new MyWebAuthnHardware();
+        String challenge = getChallenge(USERNAME, cookieFilter);
+        JsonObject registrationJson = myWebAuthnHardware.makeRegistrationJson(challenge);
+        invokeCallback(registrationJson, cookieFilter);
+        verifyLoggedIn(cookieFilter, USERNAME, User.USER);
+
+    }
+
+    @Test
+    @Order(7)
+    public void testFailLoginWithFakeRegisterUser() {
+        invokeUserLogout();
+        String newUserName = "Kipchoge";
+        ExtractableResponse<Response> response = given().filter(cookieFilter)
+                .contentType(ContentType.JSON)
+                .body("{\"name\": \"" + newUserName + "\"}")
+                .post(REGISTER_URL)
+                .then()
+                .statusCode(is(200)).extract();
+        given().filter(cookieFilter)
+                .get(PUBLIC_ME_API_URL)
+                .then()
+                .statusCode(200)
+                .body(Matchers.is("<not logged in>"));
```

Comment:
please also add call to secured endpoint that will return 401 or 403

Reply:
Well, in that test case, as the user is not logged the admin API endpoint will throw a 404, if the user is logged in correctly and the user is not an Admin the status code will be 403